### DISCUSSION TITLE
Code Error : asyncio.run() cannot be called from a running event loop

### DISCUSSION THREAD

**Discussion by Ivan-Keli** [theOehrly/Fast-F1#520] [General]
### Using the Fast-F1 signalr client?

I got this error while trying to demonstrates the usage of the SignalRClient.
Could you guys help me fix it
<img width="583" alt="Example Fast F1" src="https://github.com/theOehrly/Fast-F1/assets/116259292/0915ee74-6bb3-4df2-a38d-193eb8577fdb">

**Comment by Ivan-Keli** [comment]
### **This is what i got when i tried using ChatGPT 3.5**

The error message suggests that you are trying to run `asyncio.run()` within an already running event loop, which is not allowed. The `asyncio.run()` function is meant to be used in the context of a standalone script or application, and it creates a new event loop.

In your case, it seems like the SignalRClient is trying to start its own event loop, and there's already an existing event loop, leading to the `RuntimeError`.

To resolve this issue, you can try one of the following approaches:

1. **Use `asyncio.create_task()` instead of `asyncio.run()`:**
   Replace the line `asyncio.run(self._async_start())` with `asyncio.create_task(self._async_start())`. This will schedule the coroutine to run in the existing event loop.

   ```python
   asyncio.create_task(self._async_start())
   ```

2. **Run the event loop explicitly:**
   If you have control over the main script or application, you can run the event loop explicitly. You would remove the call to `client.start()` and add the following:

   ```python
   loop = asyncio.get_event_loop()
   loop.run_until_complete(client._async_start())
   ```

Choose one of these approaches based on your application structure and requirements.

**Comment by theOehrly** [comment]
I've converted this to issue #521. This problem is specific to running inside a Jupyter Notebook. Therefore, it also has somewhat low priority to be fixed.

**Reply by Ivan-Keli** [reply]
Thank you,
Let me go check it out
